crates/codegen/syntax_templates/src/shared/cursor.rs

```diff
@@ -0,0 +1,342 @@
+use std::rc::Rc;
+
+use super::{
+    cst::{Node, RuleNode, TokenNode},
+    kinds::{RuleKind, TokenKind},
+    text_index::{TextIndex, TextRange},
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorPathElement {
+    rule_node: Rc<RuleNode>,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorLeaf {
+    node: Node,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+impl CursorLeaf {
+    pub fn text_range(&self) -> TextRange {
+        let start = self.text_offset;
+        let end = start + self.node.text_len();
+        start..end
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Cursor {
+    path: Vec<CursorPathElement>,
+    leaf: Option<CursorLeaf>,
+}
+
+impl Cursor {
+    pub(crate) fn new(node: Node) -> Self {
+        Self {
+            path: vec![],
+            leaf: Some(CursorLeaf {
+                node,
+                child_number: 0,
+                text_offset: Default::default(),
+            }),
+        }
+    }
+
+    // Unlike clone, this re-roots at the current node.
+    // It does preserve the correct text offset however,
+    // even though the path is reset.
+    pub fn spawn(&self) -> Self {
+        Self {
+            path: vec![],
+            leaf: self.leaf.clone(),
+        }
+    }
+
+    pub fn is_finished(&self) -> bool {
+        self.leaf.is_none()
+    }
+
+    pub fn node_or_none(&self) -> Option<Node> {
```

Comment:
regarding `node()` and `node_or_none()`, I might be missing something, but I would have expected the rust idiomatic way is to return `Option<Node>` only/directly for things like this, and let the user `expect()`/`unwrap()` if needed? why the extra API?

Reply:
Because there are many cases where the client know that node isn't none e.g. after a successful `go_to_*`, and I wanted to not require `.unwrap()` there, and instead catch it in one place. Happy to change the method names if you have a suggestion?

Reply:
I think this is idiomatic to let the user make that decision, and to make it explicit at the call site (a strong direction in rust). I think similar to how a vector/map/set would offer `get() -> Option<T>` without (the often convenient) `get_and_unwrap() -> T`

Reply:
`trait Index` provides `[]` for this, and it is implemented on `Vec` and `HashMap` for example.

Reply:
But in any case, solved with an explicit completed state. Also allows the cursor to be reset